File: apps/api/plane/db/models/service_issue_allowance.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Python imports
from decimal import Decimal

# Django imports
from django.conf import settings
from django.db import models
from django.db.models import Q

# Module imports
from ..mixins import ChangeTrackerMixin
from .project import ProjectBaseModel
from .service_catalog import ServiceConfigEntity


class ServiceIssueAllowanceStatus(models.TextChoices):
    """Whether an allowance still accepts movement.

    Declared at module level, not nested, because the check constraints in ``Meta``
    reference these values and a class nested in the model is not yet in scope while
    ``Meta`` is evaluated -- the same reason ``ServiceLogSource`` and
    ``ServicePeriodStatus`` are at module level. Reachable as
    ``ServiceIssueAllowance.Status``.

    **CLOSED does not fall back to the contract pool, and that is the whole point.**
    Section 3 of the phase brief forbids the overflow of an allowance reaching the
    client's support pool "nem silenciosamente nem automaticamente"; letting a closed
    allowance route the next work log to the contract would be exactly that, arriving
    by omission instead of by decision. So a work log on a work item whose allowance is
    closed is **refused** (``ALLOWANCE_IS_CLOSED``), in the same way a closed
    competency period refuses one.
    """

    OPEN = "open", "Open"
    CLOSED = "closed", "Closed"


class ServiceIssueAllowance(ChangeTrackerMixin, ProjectBaseModel):
    """A pool of hours credited to one work item, isolated from the client's contract.

    The use case, from section 3 of the master context and the phase brief: a client on
    a 30h/month support contract closes a separate 40h project. The project gets a work
    item, the work item gets 40h of allowance, and the technicians' work logs consume
    that allowance **without touching** the monthly support pool. Rule R6 makes this
    the *first* level of the debit hierarchy, and ``plane.utils.service_pool.apply_debit``
    consults it before it consults a contract.

    **The allowance is inherited down the work item tree**, nearest ancestor first --
    see ``plane.utils.service_allowance.resolve_work_item_allowance``. A 40h project is
    broken into sub-tasks by anyone who actually executes one, and without inheritance
    the parent would hold 40h that nobody logs against while every sub-task quietly
    debited the 30h support pool. That failure is silent and it bills the wrong pool,
    which is why the consequences decided this: no inheritance costs a query, and
    inheritance costs nothing anybody would notice.

    Named ``ServiceIssueAllowance`` rather than anything built on "budget" or
    "worklog": Plane's commercial edition owns the time-tracking vocabulary and this
    fork already carries its unused hooks. Same reasoning as ``ServiceLog`` and
    ``ServiceClient``; see section 6 of the master context.

    Inherited from ``ProjectBaseModel``: ``project`` (required) and ``workspace``,
    which ``save()`` denormalises from the project. That is the base section 6 of the
    master context names for this entity by name.

    **One allowance per work item**, by partial unique index. That is not tidiness --
    it is what makes the D27 question unaskable here. D27 draws the line at *ambiguity
    blocks, absence does not*; with several allowances per work item there would be an
    ambiguous resolution to design a rule for, and with one there is nothing to
    resolve. Additional credits (acceptance criterion 5, the aditivo de escopo) are
    therefore **ledger rows on the same allowance**, never a second allowance, and the
    history of each credit with its author is the ledger.

    **``ChangeTrackerMixin`` tracks exactly one field here, and the exclusions are the
    point.** Phase 5 carried no mixin at all, with the correct reasoning that crediting
    hours is *accounting movement*: it is a ``ServiceHourLedgerEntry`` carrying its own
    ``actor`` and ``created_at``, which is the "autoria e data do crédito" the brief asks
    for. That still holds -- ``credited_hours`` and ``consumed_hours`` are accumulators
    and auditing them here would duplicate the ledger, which section 6 of the master
    context forbids; closing is recorded by ``closed_at`` and ``closed_by`` plus its own
    ledger rows; ``reference`` and ``notes`` are labels, not money.

    Phase 6 added ``overage_hour_rate``, which is **configuration that decides money**
    rather than movement, and D22 requires exactly that to be audited. So the mixin
    arrived for one field, and one only.
    """

    Status = ServiceIssueAllowanceStatus

    # Only the rate. See the class docstring for why every other column on this model is
    # deliberately absent: they are either accumulators the ledger already audits, or
    # labels that never reach an invoice.
    TRACKED_FIELDS = ["overage_hour_rate"]
    CONFIG_ENTITY_NAME = ServiceConfigEntity.ISSUE_ALLOWANCE

    # ---------------------------------------------------------------- relations

    # CASCADE, and the only foreign key on this model that is not DO_NOTHING or
    # SET_NULL. Same choice and same reasoning as `ServiceLog.issue`: an allowance has
    # no meaning apart from the work item it was credited to, so it belongs in the bin
    # alongside it -- and because `soft_delete_related_objects` handles CASCADE by soft
    # deleting, the row survives in `all_objects` for any settlement that already
    # counted it.
    issue = models.ForeignKey(
        "db.Issue",
        on_delete=models.CASCADE,
        related_name="service_allowances",
    )

    # ------------------------------------------------------------ what it is for

    # The commercial reference of the project this allowance pays for, e.g.
    # "Proposta 2026-014". Section 1 of the brief asks for it so that an allowance can
    # be tied back to the document that sold it.
    reference = models.CharField(max_length=255, blank=True)

    notes = models.TextField(blank=True)

    # ------------------------------------------------------------- the totals

    # **This row carries the balance; the ledger carries the history** -- decision D23,
    # applied unchanged. The two are written in the same transaction, always, and
    # `plane.utils.service_allowance.reconcile_allowance` asserts they still agree
    # column by column.
    #
    # Scales fixed by section 4b of the master context. Do not widen locally.

    # Every credit ever made, summed. Acceptance criteria 1 and 5: 40h then 10h leaves
    # this at 50h, and the two credits stay individually visible in the ledger.
    credited_hours = models.DecimalField(max_digits=10, decimal_places=4, default=Decimal("0.0000"))

    # The accumulator every debit moves. Always written under `select_for_update()` on
    # this row and always with an `F()` expression -- never read-modify-write.
    consumed_hours = models.DecimalField(max_digits=10, decimal_places=4, default=Decimal("0.0000"))

    # ------------------------------------------------- what happened at closing

    # Both are records of the close, not terms in the live balance. `balance_hours` is
    # `credited - consumed`; these two explain where a closed allowance's balance went,
    # and they are what makes the ledger of a closed allowance sum to exactly zero.

    # A deficit settled by billing it, in hours. Section 3: on closing an allowance in
    # deficit the Admin either credits more hours -- which is a `CREDIT` *before*
    # closing, not a settlement mode -- or bills the overage. Phase 6 converts it to
    # reais through the same mechanism `ServiceContractPeriod.overage_hours` uses: the
    # amount lands on the `OVERAGE_BILLED` ledger row, priced by `overage_hour_rate`
    # below.
    overage_hours = models.DecimalField(max_digits=10, decimal_places=4, default=Decimal("0.0000"))

    # ------------------------------------------------------------ the price of overage

    # What an hour past this allowance costs, in reais. Null falls back to the client's
    # base hour rate for the vigency in force -- see
    # `plane.utils.service_pricing.resolve_overage_rate`.
    #
    # **This column exists because the allowance is a separately negotiated sale, and
    # the support price is the wrong price for it.** A project sold at R$ 180/h does not
    # have overage at R$ 200/h because R$ 200/h is what that client's support costs; the
    # two numbers differ roughly as often as the two things are sold separately, which is
    # always. Falling back to the client's base rate as the *only* answer would price
    # project overage at the support rate and be wrong almost every time.
    #
    # It is registered now rather than later for the reason `ServiceContract
    # .overage_hour_rate` gives for itself: the moment hours are credited to a work item
    # is the moment the value of that project is known, and adding an audited column
    # afterwards means back-filling an audit trail, which cannot be done honestly. Same
    # trade the contract phase already made and accepted.
    #
    # Nullable and tracked, so it is the one field on this model that relies on the
    # `CONFIG_VALUE_UNSET` sentinel (D4) to satisfy
    # `svc_cfg_activity_shape_matches_verb`. Monetary scale (12, 2) per section 4b.
    overage_hour_rate = models.DecimalField(max_digits=12, decimal_places=2, null=True, blank=True)

    # A surplus written off at close. The project came in under budget and the hours
    # are lost, because section 3 forbids them moving to the contract pool. There is no
    # "carry" here on purpose: unlike a competency month, an allowance has no successor.
    expired_hours = models.DecimalField(max_digits=10, decimal_places=4, default=Decimal("0.0000"))

    # ------------------------------------------------------------------ lifecycle

    status = models.CharField(
        max_length=20,
        choices=ServiceIssueAllowanceStatus.choices,
        default=ServiceIssueAllowanceStatus.OPEN,
    )

    closed_at = models.DateTimeField(null=True, blank=True)

    # SET_NULL, unlike the billing foreign keys. A trace of who clicked close, not a
    # term in a calculation, following `ServiceContractPeriod.closed_by`.
    closed_by = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.SET_NULL,
        related_name="closed_service_issue_allowances",
        null=True,
        blank=True,
    )

    # Import provenance, following the convention used across plane.db.
    external_source = models.CharField(max_length=255, null=True, blank=True)
    external_id = models.CharField(max_length=255, null=True, blank=True)

    @property
    def balance_hours(self):
        """Credited minus consumed.

        A property rather than a column (decision D3), and the queryset equivalent is
        ``plane.utils.service_allowance.annotate_allowance_balance``. A
        ``GeneratedField`` was refused for the same reasons D3 gives.

        **Negative is a legitimate, expected state.** Section 3 is explicit that
        overflowing an allowance does not block work already performed -- the balance
        goes below zero and an alert fires, exactly as it does for a contract pool.
        """
        return self.credited_hours - self.consumed_hours

    @property
    def consumed_pct(self):
        """Consumption as a percentage of what was credited, or ``None`` before any credit.

        Section 4 of the brief wants the work item indicator to show "creditado,
        consumido, restante, e percentual de consumo", and the percentage is the one of
        the four that is not a subtraction.
        """
        if self.credited_hours <= 0:
            return None

        return (self.consumed_hours / self.credited_hours) * Decimal("100")

    def config_summary(self):
        """One line for the ``CREATED`` and ``DELETED`` rows of the audit trail.

        Present because ``ChangeTrackerMixin`` models carry one by convention. In
        practice only the ``UPDATED`` path is exercised for this model: an allowance row
        comes into being as a side effect of the first credit, and that event is already
        audited by the ``CREDIT`` ledger entry with its own actor -- recording it twice
        is the duplicate trail section 6 of the master context forbids.
        """
        rate = "sem valor/hora proprio" if self.overage_hour_rate is None else f"R$ {self.overage_hour_rate}/h"

        return f"Bolsa de {self.credited_hours}h ({rate})"

    class Meta:
        verbose_name = "Service Issue Allowance"
        verbose_name_plural = "Service Issue Allowances"
        db_table = "service_issue_allowances"
        ordering = ("-created_at",)

        # Uniqueness under soft delete follows the house pattern: unique_together
        # including deleted_at, plus a conditional UniqueConstraint for live rows.
        unique_together = [["issue", "deleted_at"]]

        constraints = [
            # ONE ALLOWANCE PER WORK ITEM. See the class docstring: this is what makes
            # allowance resolution total instead of ambiguous, and therefore what keeps
            # D27 from needing a second reading for this phase.
            models.UniqueConstraint(
                fields=["issue"],
                condition=Q(deleted_at__isnull=True),
                name="service_issue_allowance_unique_per_issue_when_deleted_at_null",
            ),
            # `consumed_hours` is deliberately absent from this: it is the one hour
            # column here that legitimately moves in both directions, because a
            # reversal subtracts from it.
            models.CheckConstraint(
                condition=Q(credited_hours__gte=0, overage_hours__gte=0, expired_hours__gte=0),
                name="service_issue_allowance_hours_are_not_negative",
            ),
            # An open allowance cannot have been closed by anyone, and a close with no
            # timestamp would be a decision with no date.
            models.CheckConstraint(
                condition=(
                    Q(status=ServiceIssueAllowanceStatus.OPEN, closed_at__isnull=True)
                    | Q(status=ServiceIssueAllowanceStatus.CLOSED, closed_at__isnull=False)
                ),
                name="service_issue_allowance_closed_at_matches_status",
            ),
            # Decision D20, in DDL: a rate of zero would be a second mechanism for "do
            # not charge", and this system has exactly one. Null is the legitimate "not
            # negotiated separately, use the client's base rate" state, which is why this
            # is written to permit null rather than as a plain `> 0`.
            models.CheckConstraint(
                condition=Q(overage_hour_rate__isnull=True) | Q(overage_hour_rate__gt=0),
                name="service_issue_allowance_overage_rate_is_positive",
            ),
        ]

        indexes = [
            # The debit path's first question, asked for the work item and then for
            # each ancestor up the chain.
            models.Index(fields=["issue"], name="svc_allowance_issue_idx"),
            # The workspace panel scans open allowances.
            models.Index(fields=["workspace", "status"], name="svc_allowance_ws_status_idx"),
            models.Index(fields=["project", "status"], name="svc_allowance_proj_status_idx"),
        ]

    def __str__(self):
        return f"{self.issue_id} {self.credited_hours}h"

File: packages/types/src/service-issue-allowance.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

/** Whether an allowance still accepts movement. */
export type TServiceIssueAllowanceStatus = "open" | "closed";

/**
 * A pool of hours credited to one work item, isolated from the client's contract.
 *
 * Rule R6's first level: when a work item -- or an ancestor of it -- has an allowance,
 * every `debit_pool` work log on it consumes the allowance and **never** the client's
 * monthly support pool.
 *
 * Every hour quantity is a **string**, not a number, and deliberately so, for the same
 * reason `IServiceLog` gives: the API serialises them as decimal strings and parsing
 * them here would put binary floating point into the values an invoice is built from.
 * Do not do arithmetic on them in the browser. `balance_hours` and `consumed_pct` are
 * computed server side precisely so nothing here has to.
 *
 * SECURITY, rule R11: this shape is for technicians and admins. GUEST is refused by the
 * endpoint outright; the client's own view is the portal phase, with its own payload.
 */
export interface IServiceIssueAllowance {
  readonly id: string;
  readonly issue: string;
  readonly issue_name: string | null;
  readonly project: string;

  /** The commercial reference of the project this allowance pays for. */
  readonly reference: string;
  readonly notes: string;

  readonly credited_hours: string;
  /** The same figure rendered by R3's formatter: `"50.0000"` becomes `"50h"`. Render this one. */
  readonly credited_hours_display: string;
  readonly consumed_hours: string;
  readonly consumed_hours_display: string;
  /** Credited minus consumed. **Negative is a legitimate state** -- see section 3. */
  readonly balance_hours: string;
  readonly balance_hours_display: string;
  /** Null before any credit exists: zero of zero is not zero. */
  readonly consumed_pct: string | null;

  /** A deficit billed at close, in hours. The conversion to money is a later phase. */
  readonly overage_hours: string;
  /** A surplus written off at close. It never moves to the contract pool. */
  readonly expired_hours: string;

  readonly status: TServiceIssueAllowanceStatus;
  readonly closed_at: string | null;
  readonly closed_by: string | null;

  readonly created_at: string;
  readonly updated_at: string;
}

/**
 * The four figures section 4 of the brief puts on the work item, plus the inheritance.
 *
 * `is_inherited` matters more than it looks: a technician seeing 40h on a sub-task has
 * to know those hours belong to the parent project, or the first thing they will do is
 * ask for a second allowance on the sub-task -- which is the one shape the schema
 * refuses.
 */
export interface IServiceIssueAllowanceSummary {
  readonly allowance_id: string;
  readonly issue_id: string;
  readonly reference: string;
  readonly notes: string;
  readonly status: TServiceIssueAllowanceStatus;
  /**
   * Each quantity twice: the decimal to compute with, and the `_display` twin R3's formatter
   * already rendered. The panel renders the twin; `balance_hours` stays the one that answers
   * "is this overrun", because `"−2h".startsWith("-")` is a string test on a formatted value
   * and the raw decimal is the honest place for it.
   */
  readonly credited_hours: string;
  readonly credited_hours_display: string;
  readonly consumed_hours: string;
  readonly consumed_hours_display: string;
  readonly balance_hours: string;
  readonly balance_hours_display: string;
  readonly consumed_pct: string | null;
  readonly overage_hours: string;
  readonly overage_hours_display: string;
  readonly expired_hours: string;
  readonly expired_hours_display: string;
  readonly closed_at: string | null;
  readonly is_inherited: boolean;
  readonly inherited_from_issue_id: string | null;
}

/** An allowance alert. Never a blocker -- section 3 is explicit that overrun does not block. */
export interface IServiceIssueAllowanceAlert {
  readonly code: string;
  readonly severity: string;
  readonly balance_hours?: string;
  readonly credited_hours?: string;
  readonly consumed_hours?: string;
  readonly consumed_pct?: string;
  readonly threshold_pct?: string;
}

/** One credit, as the history shows it. */
export interface IServiceIssueAllowanceCredit {
  readonly entry_id: string;
  readonly hours: string;
  readonly created_at: string;
  readonly actor_id: string | null;
  readonly actor_display_name: string | null;
  /** Set when the hours came from a contract's remaining balance. */
  readonly origin_competence: string | null;
  readonly notes: string;
}

/**
 * The response of the work item allowance endpoint.
 *
 * `allowance` is `null` when the work item has no allowance of its own and none
 * inherited. That is a legitimate answer, not an error: it means rule R6 falls through
 * to the contract pool.
 */
export interface IServiceIssueAllowanceResponse {
  readonly allowance: IServiceIssueAllowance | null;
  readonly summary?: IServiceIssueAllowanceSummary;
  readonly alerts?: IServiceIssueAllowanceAlert[];
  readonly credits?: IServiceIssueAllowanceCredit[];
}

/** What the credit form submits. Hours as a string, so the browser never rounds them. */
export interface IServiceIssueAllowanceCreditPayload {
  hours: string;
  reference?: string;
  notes?: string;
}
